File: backend/handlers/announcements.go
package handlers

import (
	"net/http"

	"ugyen-academy/db"
	"ugyen-academy/models"

	"github.com/gin-gonic/gin"
)

func GetAnnouncements(c *gin.Context) {
	user := c.MustGet("user").(models.User)

	var records []models.Announcement
	query := db.DB.Order("pinned desc, created_at desc")

	switch user.Role {
	case "student":
		query = query.Where("target_audience IN ('all', 'students')")
	case "teacher":
		query = query.Where("target_audience IN ('all', 'staff', 'teachers')")
	case "parent":
		query = query.Where("target_audience IN ('all', 'parents')")
	case "admin":
		// admin sees all
	}

	query.Find(&records)
	c.JSON(http.StatusOK, records)
}

func CreateAnnouncement(c *gin.Context) {
	user := c.MustGet("user").(models.User)

	var body struct {
		Title          string `json:"title" binding:"required"`
		Body           string `json:"body"`
		TargetAudience string `json:"targetAudience"`
		TargetClass    string `json:"targetClass"`
	}
	if err := c.ShouldBindJSON(&body); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	if body.TargetAudience == "" {
		body.TargetAudience = "all"
	}

	record := models.Announcement{
		Title:          body.Title,
		Body:           body.Body,
		TargetAudience: body.TargetAudience,
		TargetClass:    body.TargetClass,
		PostedByID:     user.ID,
	}
	if err := db.DB.Create(&record).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}
	c.JSON(http.StatusCreated, record)
}

File: backend/routes/announcements.js
const express = require('express');
const router = express.Router();
const Announcement = require('../models/Announcement');
const { protect, authorize } = require('../middleware/auth');

// @route  POST /api/announcements
// @desc   Create an announcement
// @access Private (admin, teacher)
router.post('/', protect, authorize('admin', 'teacher'), async (req, res) => {
  try {
    const announcement = await Announcement.create({
      ...req.body,
      postedBy: req.user._id
    });
    await announcement.populate('postedBy', 'name email role');
    res.status(201).json({ success: true, data: announcement });
  } catch (err) {
    console.error(err);
    res.status(500).json({ success: false, message: err.message || 'Server Error' });
  }
});

// @route  GET /api/announcements
// @desc   Get announcements relevant to the current user's role
// @access Private
router.get('/', protect, async (req, res) => {
  try {
    const role = req.user.role;

    // Build audience filter
    const audienceFilter = ['all'];
    if (role === 'teacher') audienceFilter.push('teachers');
    if (role === 'student') audienceFilter.push('students');
    if (role === 'parent') audienceFilter.push('parents');
    if (role === 'admin') {
      audienceFilter.push('teachers', 'students', 'parents', 'class');
    }

    const query = {
      targetAudience: { $in: audienceFilter },
      $or: [{ expiresAt: { $exists: false } }, { expiresAt: null }, { expiresAt: { $gte: new Date() } }]
    };

    const announcements = await Announcement.find(query)
      .populate('postedBy', 'name email role')
      .sort({ pinned: -1, createdAt: -1 });

    res.json({ success: true, data: announcements });
  } catch (err) {
    console.error(err);
    res.status(500).json({ success: false, message: err.message || 'Server Error' });
  }
});

// @route  PUT /api/announcements/:id
// @desc   Update an announcement (own, unless admin)
// @access Private (admin, teacher)
router.put('/:id', protect, authorize('admin', 'teacher'), async (req, res) => {
  try {
    let announcement = await Announcement.findById(req.params.id);
    if (!announcement) {
      return res.status(404).json({ success: false, message: 'Announcement not found' });
    }

    // Teachers can only update their own announcements
    if (
      req.user.role === 'teacher' &&
      announcement.postedBy.toString() !== req.user._id.toString()
    ) {
      return res.status(403).json({ success: false, message: 'Not authorized to update this announcement' });
    }

    announcement = await Announcement.findByIdAndUpdate(req.params.id, req.body, {
      new: true,
      runValidators: true
    }).populate('postedBy', 'name email role');

    res.json({ success: true, data: announcement });
  } catch (err) {
    console.error(err);
    res.status(500).json({ success: false, message: err.message || 'Server Error' });
  }
});

// @route  DELETE /api/announcements/:id
// @desc   Delete an announcement
// @access Private (admin)
router.delete('/:id', protect, authorize('admin'), async (req, res) => {
  try {
    const announcement = await Announcement.findByIdAndDelete(req.params.id);
    if (!announcement) {
      return res.status(404).json({ success: false, message: 'Announcement not found' });
    }
    res.json({ success: true, data: {} });
  } catch (err) {
    console.error(err);
    res.status(500).json({ success: false, message: err.message || 'Server Error' });
  }
});

module.exports = router;
